src/transport/SecureSessionMgr.cpp

```diff
@@ -111,54 +120,74 @@ CHIP_ERROR SecureSessionMgr::SendMessage(PayloadHeader & payloadHeader, NodeId p
 
     {
         uint8_t * data = nullptr;
+        uint8_t * p    = nullptr;
+        uint16_t len   = 0;
         PacketHeader packetHeader;
         MessageAuthenticationCode mac;
 
         const uint16_t headerSize = payloadHeader.EncodeSizeBytes();
         uint16_t actualEncodedHeaderSize;
         uint16_t totalLen = 0;
         uint16_t taglen   = 0;
-        uint32_t payloadLength; // Make sure it's big enough to add two 16-bit
-                                // ints without overflowing.
-        static_assert(std::is_same<decltype(msgBuf->TotalLength()), uint16_t>::value,
-                      "Addition to generate payloadLength might overflow");
-        payloadLength = static_cast<uint32_t>(headerSize + msgBuf->TotalLength());
-        VerifyOrExit(CanCastTo<uint16_t>(payloadLength), err = CHIP_ERROR_NO_MEMORY);
+
+        if (!isResend)
+        {
+            msgId = state->GetSendMessageIndex();
+
+            static_assert(std::is_same<decltype(msgBuf->TotalLength()), uint16_t>::value,
+                          "Addition to generate payloadLength might overflow");
+            payloadLength = static_cast<uint32_t>(headerSize + msgBuf->TotalLength());
+            VerifyOrExit(CanCastTo<uint16_t>(payloadLength), err = CHIP_ERROR_NO_MEMORY);
+        }
 
         packetHeader
             .SetSourceNodeId(mLocalNodeId)              //
             .SetDestinationNodeId(peerNodeId)           //
-            .SetMessageId(state->GetSendMessageIndex()) //
+            .SetMessageId(msgId)                        //
             .SetEncryptionKeyID(state->GetLocalKeyID()) //
             .SetPayloadLength(static_cast<uint16_t>(payloadLength));
         packetHeader.GetFlags().Set(Header::FlagValues::kSecure);
 
         ChipLogProgress(Inet, "Sending msg from %llu to %llu", mLocalNodeId, peerNodeId);
 
-        VerifyOrExit(msgBuf->EnsureReservedSize(headerSize), err = CHIP_ERROR_NO_MEMORY);
+        // Skip encryption process if the packet is resent by CRMP
+        if (!isResend)
```

Comment:
This change is adding a mechanism to bypass the encryption of messages in the secure channel, which is a big security hole. So far it was guaranteed that any message sent thru SecureSessionMgr is encrypted. The user of the `SendMessage()` can simply set `isResend` to true to bypass encryption.

Reply:
Good catch, made the overloaded SendMessage() which can overpass the encryption process private, this should prevent application from bypassing the encryption of messages in the secure channel.